src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb

```diff
@@ -303,53 +303,32 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-                     SyntaxKind.EqualsKeyword,
-                     SyntaxKind.AscendingKeyword,
-                     SyntaxKind.DescendingKeyword,
-                     SyntaxKind.WhereKeyword
-                    Return True
-                Case SyntaxKind.InKeyword
-                    Select Case token.Parent.Kind()
-                        Case SyntaxKind.FromClause,
-                             SyntaxKind.SimpleJoinClause,
-                             SyntaxKind.GroupJoinClause
-                            Return True
-                        Case Else
-                            Return False ' e.g. ForEach ... in
-                    End Select
+                Case SyntaxKind.JoinKeyword
+                Case SyntaxKind.IntoKeyword
+                Case SyntaxKind.AggregateKeyword
+                Case SyntaxKind.DistinctKeyword
+                Case SyntaxKind.SkipKeyword
+                Case SyntaxKind.TakeKeyword
+                Case SyntaxKind.LetKeyword
+                Case SyntaxKind.ByKeyword
+                Case SyntaxKind.OrderKeyword
+                Case SyntaxKind.WhereKeyword
                 Case SyntaxKind.OnKeyword
-                    Select Case token.Parent.Kind()
-                        Case SyntaxKind.SimpleJoinClause,
-                             SyntaxKind.GroupJoinClause
-                            Return True
-                        Case Else
-                            Return False ' e.g. On Error Goto
-                    End Select
                 Case SyntaxKind.FromKeyword
-                    Return token.Parent.Kind() = SyntaxKind.FromClause ' False for e.g. Object collection initializer
                 Case SyntaxKind.WhileKeyword
-                    Select Case token.Parent.Kind()
-                        Case SyntaxKind.SkipWhileClause,
-                             SyntaxKind.TakeWhileClause
-                            Return True
-                        Case Else
-                            Return False ' e.g. Exit While
-                    End Select
                 Case SyntaxKind.SelectKeyword
-                    Return token.Parent.Kind() = SyntaxKind.SelectClause ' False for e.g. Select Case
-                Case Else
-                    Return False
+                    Return TypeOf token.Parent Is QueryClauseSyntax
+                Case SyntaxKind.GroupKeyword
+                    Return (TypeOf token.Parent Is QueryClauseSyntax) OrElse (token.Parent.IsKind(SyntaxKind.GroupAggregation))
+                Case SyntaxKind.EqualsKeyword
+                    Return TypeOf token.Parent Is JoinConditionSyntax
+                Case SyntaxKind.AscendingKeyword
+                Case SyntaxKind.DescendingKeyword
+                    Return TypeOf token.Parent Is OrderingSyntax
+                Case SyntaxKind.InKeyword
+                    Return TypeOf token.Parent Is CollectionRangeVariableSyntax
             End Select
+            Return False
```

Comment:
Put this inside "Case Else" to give some automated tools a chance to catch your fall-through bug.